NebulaWorld/Player/DroneManager.cs

```diff
@@ -192,181 +77,215 @@ public static void RefreshCachedPositions()
+        for (var droneId = 1; droneId < drones.cursor; droneId++)
         {
-            nearestPlayer = 1; // this is host
-            shortestDistance = (GameMain.mainPlayer.position - entityPos).sqrMagnitude;
-        }
+            ref var ptr = ref drones.buffer[droneId];
+            if (ptr.owner != planetId) //ptr.owner is planetId in the custom pool
+            {
+                continue;
+            }
+            var craftData = crafts[ptr.craftId];
+            var playerId = (ushort)craftData.owner;
+            if (!cachedPositions.TryGetValue(playerId, out var playerPosition) || playerPosition.PlanetId != planetId)
+            {
+                // If the owner leave the planet, recycle the drone
+                RecycleDrone(factory, ref ptr, droneId);
+                continue;
+            }
 
-        foreach (var playerPosition in CachedPositions.Where(playerPosition =>
-                     playerPosition.Value.PlanetId == planetId))
-        {
-            var dist = (playerPosition.Value.Position - entityPos).sqrMagnitude;
+            // Update drone stage and craft position
+            var flag2 = false;
+            var ejectPos = playerPosition.Position;
+            var meachEnerey = (double)float.MaxValue; // Dummy value for remote drones
+            var mecahEnergyChange = 0.0;
+            var result = ptr.InternalUpdate(ref craftData, factory, ref ejectPos, constructionDroneSpeed, dt,
+                ref meachEnerey, ref mecahEnergyChange, 0, 0, out _);
+            crafts[ptr.craftId] = craftData;
 
-            if (nearestPlayer == 0 && dist <= buildArea)
+            if (ptr.stage == 3) // Hovering state?
             {
-                nearestPlayer = playerPosition.Key;
-                shortestDistance = dist;
-
-                continue;
+                if (ptr.targetObjectId > 0) // Repair entity
+                {
+                    result = (factory.constructionSystem.Repair(ptr.targetObjectId, 1.0f, time) ? 1 : 0); // Assume energy ratio is 1.0f
+                    if (result == 1)
+                    {
+                        ptr.targetObjectId = 0;
+                    }
+                }
+                else if (result == 0) // Mod: Do not wait for prebuild to finished, just advance to the next target
+                {
+                    result = 1;
+                }
+            }
+            if (result == 1 && ptr.stage == 4)
+            {
+                RecycleDrone(factory, ref ptr, droneId);
+            }
+            if (result != 0 && (ptr.stage == 2 || ptr.stage == 3 || ptr.stage == 4))
+            {
+                ptr.movement--;
+                if (ptr.movement <= 0 || flag2)
+                {
+                    ptr.movement = 0;
+                    ptr.stage = 4;
+                    ptr.targetObjectId = 0;
+                }
+                else if (ptr.nextTarget1ObjectId != 0)
```

Comment:
This really has a refactor me feel, but I dont have the full context. Seems like the entire conditional set belongs in a method. 

Also, the 1, 2 ,3 seems like you are either following DSP values, or 1,2, 3 are specifically coded. Hard to tell withotu a deep dive.

Reply:
As the comment state, this function is a modified version of `ConstructionSystem.UpdateDrones()`.
Those 1,2,3 are DSP values used in `DroneComponent.InternalUpdate()`.